### PR TITLE
feature(xcy): add ReZero algo.

### PR DESCRIPTION
**Pull request by HarryXuancy**
- Our work is currently in progress, focusing on developing an efficient version of MuZero.
- Code are submitted for discussion.
- Our arxiv paper is on [ReZero: Boosting MCTS-based Algorithms by Backward-view and Entire-buffer Reanalyze](https://arxiv.org/abs/2404.16364).

### CODE REVIEW

File: lzero/mcts/buffer/game_buffer_ma.py [hunk 1]
```diff
@@ -571,17 +716,265 @@ def _compute_target_policy_reanalyzed(self, policy_re_context: List[Any], model:
...
+
+            # 根据得到的这些latent 根节点搜索新的target_policy
+            # prepare之后构建新的roots类，并把现有的roots分批复制进去
+            # if self._cfg.mcts_ctree:
+            #     # cpp mcts_tree
+            #     roots = MCTSCtree.roots(transition_batch_size, legal_actions)
+            #     roots.prepare(self._cfg.root_noise_weight, noises, reward_pool, policy_logits_pool, to_play)
+            #     # do MCTS for a new policy with the recent target model
+            #     MCTSCtree(self._cfg).search(roots, model, latent_state_roots, to_play)
+            if self._cfg.mcts_ctree:
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer_ma.py:855]: def compute_all_filters(data, num_unroll_steps):
    data_by_iter = []
    for iter in range(num_unroll_steps + 1):
        iter_data = [x for i, x in enumerate(data)
                     if (i + 1) % (num_unroll_steps + 1) ==
                     ((num_unroll_steps + 1 - iter) % (num_unroll_steps + 1))]
        data_by_iter.append(iter_data)
    return data_by_iter

if self._cfg.mcts_ctree:
    legal_actions_by_iter = compute_all_filters(legal_actions, self._cfg.num_unroll_steps)
    noises_by_iter = compute_all_filters(noises, self._cfg.num_unroll_steps)
    reward_pool_by_iter = compute_all_filters(reward_pool, self._cfg.num_unroll_steps)
    policy_logits_pool_by_iter = compute_all_filters(policy_logits_pool, self._cfg.num_unroll_steps)
    to_play_by_iter = compute_all_filters(to_play, self._cfg.num_unroll_steps)
    latent_state_roots_by_iter = compute_all_filters(latent_state_roots, self._cfg.num_unroll_steps)
    true_action_by_iter = compute_all_filters(true_action, self._cfg.num_unroll_steps)

    temp_values = []
    temp_distributions = []
    mcts_ctree = MCTSCtree(self._cfg)

    for iter in range(self._cfg.num_unroll_steps + 1):
        iter_batch_size = transition_batch_size / (self._cfg.num_unroll_steps + 1)
        roots = MCTSCtree.roots(iter_batch_size, legal_actions_by_iter[iter])

        roots.prepare(self._cfg.root_noise_weight, 
                      noises_by_iter[iter], 
                      reward_pool_by_iter[iter],
                      policy_logits_pool_by_iter[iter], 
                      to_play_by_iter[iter])

        if iter == 0:
            mcts_ctree.search(roots, model, latent_state_roots_by_iter[iter], to_play_by_iter[iter])
        else:
            mcts_ctree.search_with_reuse(roots, model, latent_state_roots_by_iter[iter], 
                                         to_play_by_iter[iter],
                                         true_action_list=true_action_by_iter[iter], 
                                         reuse_value_list=iter_values)

        iter_values = roots.get_values()
        iter_distributions = roots.get_distributions()
        temp_values.append(iter_values)
        temp_distributions.append(iter_distributions)

File: lzero/entry/__init__.py [hunk 2]
```diff
@@ -4,4 +4,9 @@
 from .train_muzero_with_reward_model import train_muzero_with_reward_model
 from .eval_muzero import eval_muzero
 from .eval_muzero_with_gym_env import eval_muzero_with_gym_env
-from .train_muzero_with_gym_env import train_muzero_with_gym_env
\ No newline at end of file
+from .train_muzero_with_gym_env import train_muzero_with_gym_env
```
Comment (puyuan1996) [inline lzero/entry/__init__.py:7]: 不同的entry都尽可能整合到train_muzero.py中，通过传入一个option来选择具体的不同处理方式。与train_muzero.py差别很大的才单独新建文件。尽可能保持代码的整洁和可维护性。

File: lzero/mcts/buffer/game_buffer.py [hunk 3]
```diff
@@ -148,6 +150,50 @@ def _sample_orig_data(self, batch_size: int) -> Tuple:
 
         orig_data = (game_segment_list, pos_in_game_segment_list, batch_index_list, weights_list, make_time)
         return orig_data
+    
+    def _sample_orig_reanalyze_data(self, batch_size: int) -> Tuple:
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer.py:154]: 这个新增方法的目的是？

File: lzero/mcts/buffer/game_buffer.py [hunk 4]
```diff
@@ -148,6 +150,50 @@ def _sample_orig_data(self, batch_size: int) -> Tuple:
...
+        
+            batch_index_list = np.random.choice(2000, batch_size, replace=False)
+
+
+            if self._cfg.reanalyze_outdated is True:
+                # NOTE: used in reanalyze part
+                batch_index_list.sort()
+
+            # weights_list = (num_of_transitions * probs[batch_index_list]) ** (-self._beta)
+            # weights_list /= weights_list.max()
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer.py:180]: 把目前用不到的注释可以删除了哈，如果是TODO需要后续check的写清楚注释

File: lzero/mcts/buffer/game_buffer_ma.py [hunk 5]
```diff
@@ -0,0 +1,1339 @@
+from typing import Any, List, Tuple, Union, TYPE_CHECKING, Optional
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer_ma.py:1]: game_buffer也类似，通过继承之前的game_buffer，只增加最必须的方法，其他的尽可能复用哈，不需要的中文注释删除了，必要的可以转变成扼要的英文注释哈

File: lzero/mcts/ctree/ctree_efficientzero/lib/cnode.cpp [hunk 6]
```diff
@@ -600,6 +600,90 @@ namespace tree
         }
     }
 
+    void cbatch_backpropagate_with_reuse(int current_latent_state_index, float discount_factor, const std::vector<float> &value_prefixs, const std::vector<float> &values, const std::vector<std::vector<float> > &policies, tools::CMinMaxStatsList *min_max_stats_lst, CSearchResults &results, std::vector<int> is_reset_list, std::vector<int> &to_play_batch, std::vector<int> &no_inference_lst, std::vector<int> &reuse_lst, std::vector<float> &reuse_value_lst)
+    {
+        /*
+        Overview:
+            Expand the nodes along the search path and update the infos.
```
Comment (puyuan1996) [inline lzero/mcts/ctree/ctree_efficientzero/lib/cnode.cpp:607]: ctree中增加的方法，注释也对应修改一下哈

File: lzero/mcts/ptree/ptree_mz.py [hunk 7]
```diff
@@ -418,6 +463,60 @@ def compute_ucb_score(
...
+    pb_c = math.log((total_children_visit_counts + pb_c_base + 1) / pb_c_base) + pb_c_init
+    pb_c *= (math.sqrt(total_children_visit_counts) / (child.visit_count + 1))
+    prior_score = pb_c * child.prior
+
+    # Compute the value score.
+    if child.visit_count == 0:
+        value_score = parent_mean_q
+    else:
+        # !!!!!!!!!!!!!!!!!!!这边暂时不考虑玩家切换的问题！！！！！！！！！！！！！！！！！！！！！！！！1
+        true_reward = child.reward
```
Comment (puyuan1996) [inline lzero/mcts/ptree/ptree_mz.py:502]: ptree中也是哈，注释修改一下。对了目前rezero的代码都是用ctree跑的吧，ptree是最开始为了原理调试写的吧

File: lzero/mcts/tree_search/mcts_ctree.py [hunk 8]
```diff
@@ -390,6 +534,169 @@ def search(
...
 
+    def search_with_reuse(
+            self, roots: Any, model: torch.nn.Module, latent_state_roots: List[Any],
+            reward_hidden_state_roots: List[Any], to_play_batch: Union[int, List[Any]],
+            true_action_list = None,
+            reuse_value_list = None
+    ) -> None:
+        """
+        Overview:
+            Do MCTS for the roots (a batch of root nodes in parallel). Parallel in model inference.
```
Comment (puyuan1996) [inline lzero/mcts/tree_search/mcts_ctree.py:545]: 更新注释

File: lzero/policy/ma.py [hunk 9]
```diff
@@ -0,0 +1,838 @@
...
+from lzero.model import ImageTransforms
+from lzero.policy import scalar_transform, InverseScalarTransform, cross_entropy_loss, phi_transform, \
+    DiscreteSupport, to_torch_float_tensor, mz_network_output_unpack, select_action, negative_cosine_similarity, \
+    prepare_obs
+
+
+@POLICY_REGISTRY.register('ma')
+class MAPolicy(Policy):
+    """
+    Overview:
```
Comment (puyuan1996) [inline lzero/policy/ma.py:25]: 改成继承自muzero，并重新命名为 rezero ，只增加修改后的方法哈，尽可能保持代码整洁易懂

File: lzero/worker/ma_collector.py [hunk 10]
```diff
@@ -0,0 +1,773 @@
...
+
+from lzero.mcts.buffer.game_segment import GameSegment
+from lzero.mcts.utils import prepare_observation
+
+
+@SERIAL_COLLECTOR_REGISTRY.register('episode_ma')
+class MACollector(ISerialCollector):
+    """
+    Overview:
+        The Collector for MCTS+RL algorithms, including MuZero, EfficientZero, Sampled EfficientZero, Gumbel MuZero.
```
Comment (puyuan1996) [inline lzero/worker/ma_collector.py:22]: 同样的命名，注释等改写一下，继承自muzero_collector

File: zoo/atari/config/main_config/pong_mcma.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
+from easydict import EasyDict
+
```
Comment (puyuan1996) [inline zoo/atari/config/main_config/pong_mcma.py:2]: atari的所有config，对于一个算法，不同环境公用同一个config，里面用分支选择特定的参数，config对应的算法命名改一下。

**Review by puyuan1996 — CHANGES_REQUESTED** [review]

File: Untitled Diagram.drawio [hunk 12]
```diff
@@ -0,0 +1,10 @@
+<mxfile host="app.diagrams.net" modified="2024-01-24T07:16:11.105Z" agent="Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:88.0) Gecko/20100101 Firefox/88.0" version="21.8.2" etag="4qZMPrcOvV_-34SzBdEL" type="github">
```
[the remark this answers is not in the document]
Comment (HarryXuancy) [reply Untitled Diagram.drawio:1]: 好的这个我删除一下

File: lzero/mcts/buffer/game_buffer.py (hunk 3 above)
Comment (HarryXuancy) [reply lzero/mcts/buffer/game_buffer.py:154]: 这个应该是在做buffer reanalyze的时候切分buffer用的，因为reanalyze的时候是2000个样本做一个batch一起搜索，所以要通过sample把样本组织成一个个batch，这一块的方法我后续可能再优化一下

File: lzero/mcts/ptree/ptree_mz.py (hunk 7 above)
Comment (HarryXuancy) [reply lzero/mcts/ptree/ptree_mz.py:502]: 是的，ptree当时改了一半，再后来就没有调试过了，暂时还没有更新ptree版本的代码，这些注释我删一下

File: lzero/entry/train_nrma.py [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        weight_temp = weights_list[i*batch_size:(i+1)*batch_size]
+        time_temp = make_time_list[i*batch_size:(i+1)*batch_size]
+        current_temp = [obs_temp, action_temp, mask_temp, index_temp, weight_temp, time_temp]
+        iter_data.append([current_temp, target_temp])
+    
+    return iter_data
+    
+
+def train_nrma(
+        input_cfg: Tuple[dict, dict],
```
Comment (puyuan1996) [inline lzero/entry/train_nrma.py:54]: 这个函数名改成更详细一点的吧，希望能见名知意。

File: lzero/entry/train_nrma.py [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        seed: int = 0,
+        model: Optional[torch.nn.Module] = None,
+        model_path: Optional[str] = None,
+        max_train_iter: Optional[int] = int(1e10),
+        max_env_step: Optional[int] = int(1e10),
+) -> 'Policy':  # noqa
+    """
```
Comment (puyuan1996) [inline lzero/entry/train_nrma.py:61]: 改一下注释

File: lzero/entry/train_nrma.py [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        for i in range(cfg.policy.buffer_reanalyze_freq):
+            if replay_buffer.get_num_of_transitions()>2000 and reanalyze_count%2 == 0:
+                replay_buffer.reanalyze_buffer(2000, policy)
+
+            for i in range(update_per_collect):
+                # Learner will train ``update_per_collect`` times in one iteration.
+                if replay_buffer.get_num_of_transitions() > batch_size:
+
+
+                    # policy对sample的影响是什么？？？？？？？？？？？？？？？？？？？？？？？？？
```
Comment (puyuan1996) [inline lzero/entry/train_nrma.py:233]: 多余注释删了，必要的注释和TODO，用英文简要地写一下？

File: zoo/atari/config/main_config/asterix_ma.py [hunk 16]
```diff
@@ -0,0 +1,116 @@
+from easydict import EasyDict
```
Comment (puyuan1996) [inline zoo/atari/config/main_config/asterix_ma.py:1]: 类似这样 https://github.com/opendilab/LightZero/blob/dev-unizero/zoo/atari/config/atari_unizero_config_stack1_atari20games.py 将atari的各个算法的config都整合一下，每个算法只需要一个config即可

File: lzero/worker/ma_collector.py [hunk 17]
```diff
@@ -0,0 +1,773 @@
...
+                        continue
+                    obs, reward, done, info = timestep.obs, timestep.reward, timestep.done, timestep.info
+
+                    # if self.policy_config.sampled_algo:
+                    #     game_segments[env_id].store_search_stats(
+                    #         distributions_dict[env_id], value_dict[env_id], root_sampled_actions_dict[env_id]
+                    #     )
+                    # elif self.policy_config.gumbel_algo:
+                    #     game_segments[env_id].store_search_stats(distributions_dict[env_id], value_dict[env_id], improved_policy = improved_policy_dict[env_id])
+                    # else:
```
Comment (puyuan1996) [inline lzero/worker/ma_collector.py:503]: 注释整合修改

File: lzero/mcts/buffer/game_buffer_nrma.py [hunk 18]
```diff
@@ -0,0 +1,1337 @@
...
+                    if (i + 1) % (num_unroll_steps + 1) ==
+                    ((num_unroll_steps + 1 - iter) % (num_unroll_steps + 1))]
+        data_by_iter.append(iter_data)
+    return data_by_iter
+    
+@BUFFER_REGISTRY.register('game_buffer_nrma')
+class NRMAGameBuffer(GameBuffer):
+    """
+    Overview:
+        The specific game buffer for MuZero policy.
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer_nrma.py:33]: 命名和注释改一下吧。nrma读者很难知道具体的功能

File: lzero/mcts/buffer/game_buffer_rezeroe.py [hunk 19]
```diff
@@ -0,0 +1,737 @@
...
+def compute_all_filters(data, num_unroll_steps):
+    data_by_iter = []
+    for iter in range(num_unroll_steps + 1):
+        iter_data = [x for i, x in enumerate(data)
+                    if (i + 1) % (num_unroll_steps + 1) ==
+                    ((num_unroll_steps + 1 - iter) % (num_unroll_steps + 1))]
+        data_by_iter.append(iter_data)
+    return data_by_iter
+    
+@BUFFER_REGISTRY.register('game_buffer_rezeroe')
```
Comment (puyuan1996) [inline lzero/mcts/buffer/game_buffer_rezeroe.py:29]: 或许改成game_buffer_rezero_ez更加清晰一些哈

**Comment by puyuan1996** [comment]
We have a new polished PR https://github.com/opendilab/LightZero/pull/238.

**Pull request closed without merging**